File: bindings/php/Palette.php
<?php
namespace Stencil;

final class Palette {
    /** @internal */
    public \FFI\CData $ptr;
    private bool $owned;
    private ?object $borrowedFrom;

    /** @internal */
    public function __construct(\FFI\CData $ptr, bool $owned, ?object $borrowedFrom = null) {
        $this->ptr = $ptr;
        $this->owned = $owned;
        $this->borrowedFrom = $borrowedFrom;
    }

    public static function all() {
        $ret = Lib::ffi()->Palette_all();
        return new Palette($ret, true);
    }

    public static function solid() {
        $ret = Lib::ffi()->Palette_solid();
        return new Palette($ret, true);
    }

    public static function structural() {
        $ret = Lib::ffi()->Palette_structural();
        return new Palette($ret, true);
    }

    public static function decorative() {
        $ret = Lib::ffi()->Palette_decorative();
        return new Palette($ret, true);
    }

    public static function concrete() {
        $ret = Lib::ffi()->Palette_concrete();
        return new Palette($ret, true);
    }

    public static function wool() {
        $ret = Lib::ffi()->Palette_wool();
        return new Palette($ret, true);
    }

    public static function terracotta() {
        $ret = Lib::ffi()->Palette_terracotta();
        return new Palette($ret, true);
    }

    public static function grayscale() {
        $ret = Lib::ffi()->Palette_grayscale();
        return new Palette($ret, true);
    }

    public static function wood() {
        $ret = Lib::ffi()->Palette_wood();
        return new Palette($ret, true);
    }

    public function dithered() {
        $ret = Lib::ffi()->Palette_dithered($this->ptr);
        return new Palette($ret, true);
    }

    public function sortedByLightness() {
        $ret = Lib::ffi()->Palette_sorted_by_lightness($this->ptr);
        return new Palette($ret, true);
    }

    public function rampIdsJson( $r1,  $g1,  $b1,  $r2,  $g2,  $b2,  $steps) {
        $write = Lib::ffi()->diplomat_buffer_write_create(0);
        $result = Lib::ffi()->Palette_ramp_ids_json($this->ptr, $r1, $g1, $b1, $r2, $g2, $b2, $steps, $write);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
        return Lib::readAndFreeWrite($write);
    }

    public function gradientIdsJson( $r1,  $g1,  $b1,  $r2,  $g2,  $b2,  $steps) {
        $write = Lib::ffi()->diplomat_buffer_write_create(0);
        $result = Lib::ffi()->Palette_gradient_ids_json($this->ptr, $r1, $g1, $b1, $r2, $g2, $b2, $steps, $write);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
        return Lib::readAndFreeWrite($write);
    }

    public static function fromBlockIds(string $ids_json) {
        $__n0 = strlen($ids_json);
        $__view0 = Lib::ffi()->new('DiplomatStringView');
        if ($__n0 > 0) {
            $__buf0 = Lib::ffi()->new("uint8_t[" . $__n0 . "]", false);
            \FFI::memcpy($__buf0, $ids_json, $__n0);
            $__view0->data = Lib::ffi()->cast('const char*', \FFI::addr($__buf0[0]));
        } else {
            $__view0->data = null;
        }
        $__view0->len = $__n0;
        $result = Lib::ffi()->Palette_from_block_ids($__view0);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
        return new Palette($result->ok, true);
    }

    public function len() {
        $ret = Lib::ffi()->Palette_len($this->ptr);
        return $ret;
    }

    public function blockIdsJson() {
        $write = Lib::ffi()->diplomat_buffer_write_create(0);
        Lib::ffi()->Palette_block_ids_json($this->ptr, $write);
        return Lib::readAndFreeWrite($write);
    }

    public function closestBlockDithered( $r,  $g,  $b,  $x,  $y,  $z) {
        $write = Lib::ffi()->diplomat_buffer_write_create(0);
        $result = Lib::ffi()->Palette_closest_block_dithered($this->ptr, $r, $g, $b, $x, $y, $z, $write);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
        return Lib::readAndFreeWrite($write);
    }

    public function closestBlock( $r,  $g,  $b) {
        $write = Lib::ffi()->diplomat_buffer_write_create(0);
        $result = Lib::ffi()->Palette_closest_block($this->ptr, $r, $g, $b, $write);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
        return Lib::readAndFreeWrite($write);
    }

    public function __destruct() {
        if ($this->owned) {
            Lib::ffi()->Palette_destroy($this->ptr);
        }
    }
}

File: bindings/kotlin/src/main/kotlin/at/schem/nucleation/Palette.kt
package at.schem.nucleation;
import com.sun.jna.Callback
import com.sun.jna.Library
import com.sun.jna.Native
import com.sun.jna.Pointer
import com.sun.jna.Structure

internal interface PaletteLib: Library {
    fun Palette_destroy(handle: Pointer)
    fun Palette_all(): Pointer
    fun Palette_solid(): Pointer
    fun Palette_structural(): Pointer
    fun Palette_decorative(): Pointer
    fun Palette_concrete(): Pointer
    fun Palette_wool(): Pointer
    fun Palette_terracotta(): Pointer
    fun Palette_grayscale(): Pointer
    fun Palette_wood(): Pointer
    fun Palette_sorted_by_lightness(handle: Pointer): Pointer
    fun Palette_ramp_ids_json(handle: Pointer, r1: FFIUint8, g1: FFIUint8, b1: FFIUint8, r2: FFIUint8, g2: FFIUint8, b2: FFIUint8, steps: FFIUint32, write: Pointer): ResultUnitInt
    fun Palette_gradient_ids_json(handle: Pointer, r1: FFIUint8, g1: FFIUint8, b1: FFIUint8, r2: FFIUint8, g2: FFIUint8, b2: FFIUint8, steps: FFIUint32, write: Pointer): ResultUnitInt
    fun Palette_from_block_ids(idsJson: Slice): ResultPointerInt
    fun Palette_len(handle: Pointer): FFISizet
    fun Palette_block_ids_json(handle: Pointer, write: Pointer): Unit
    fun Palette_closest_block(handle: Pointer, r: FFIUint8, g: FFIUint8, b: FFIUint8, write: Pointer): ResultUnitInt
}
/** A set of colored blocks that color/gradient brushes snap their computed
*colors to (nearest neighbor in Oklab space). Wraps an Arc'd
*[crate::building::BlockPalette]; sharing one palette across many
*brushes is cheap.
*/
class Palette internal constructor (
    internal val handle: Pointer,
    // These ensure that anything that is borrowed is kept alive and not cleaned
    // up by the garbage collector.
    internal val selfEdges: List<Any>,
    internal var owned: Boolean,
)  {

    init {
        if (this.owned) {
            this.registerCleaner()
        }
    }

    private class PaletteCleaner(val handle: Pointer, val lib: PaletteLib) : Runnable {
        override fun run() {
            lib.Palette_destroy(handle)
        }
    }
    private fun registerCleaner() {
        CLEANER.register(this, Palette.PaletteCleaner(handle, Palette.lib));
    }

    companion object {
        internal val libClass: Class<PaletteLib> = PaletteLib::class.java
        internal val lib: PaletteLib = Native.load("nucleation", libClass)
        @JvmStatic
        
        /** Every block blockpedia knows a color for (the default palette
        *brushes use when none is set).
        */
        fun all(): Palette {
            
            val returnVal = lib.Palette_all();
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = Palette(handle, selfEdges, true)
            return returnOpaque
        }
        @JvmStatic
        
        /** Only solid blocks: no transparency, gravity, tile entities, or
        *support requirements.
        */
        fun solid(): Palette {
            
            val returnVal = lib.Palette_solid();
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = Palette(handle, selfEdges, true)
            return returnOpaque
        }
        @JvmStatic
        
        /** Conservative structural set (full building blocks).
        */
        fun structural(): Palette {
            
            val returnVal = lib.Palette_structural();
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = Palette(handle, selfEdges, true)
            return returnOpaque
        }
        @JvmStatic
        
        /** Decorative set: allows stairs/slabs but no tile entities.
        */
        fun decorative(): Palette {
            
            val returnVal = lib.Palette_decorative();
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = Palette(handle, selfEdges, true)
            return returnOpaque
        }
        @JvmStatic
        
        /** The 16 concrete colors (excludes concrete powder).
        */
        fun concrete(): Palette {
            
            val returnVal = lib.Palette_concrete();
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = Palette(handle, selfEdges, true)
            return returnOpaque
        }
        @JvmStatic
        
        /** The 16 wool colors.
        */
        fun wool(): Palette {
            
            val returnVal = lib.Palette_wool();
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = Palette(handle, selfEdges, true)
            return returnOpaque
        }
        @JvmStatic
        
        /** Terracotta colors (excludes glazed variants).
        */
        fun terracotta(): Palette {
            
            val returnVal = lib.Palette_terracotta();
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = Palette(handle, selfEdges, true)
            return returnOpaque
        }
        @JvmStatic
        
        /** Genuinely gray blocks: opaque full cubes whose measured color
        *is near-neutral (low Oklab chroma) — judged from color data,
        *not names.
        */
        fun grayscale(): Palette {
            
            val returnVal = lib.Palette_grayscale();
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = Palette(handle, selfEdges, true)
            return returnOpaque
        }
        @JvmStatic
        
        /** The planks family — a natural light→dark wood ramp.
        */
        fun wood(): Palette {
            
            val returnVal = lib.Palette_wood();
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = Palette(handle, selfEdges, true)
            return returnOpaque
        }
        @JvmStatic
        
        /** Custom palette from a JSON array of block ids, e.g.
        *`["minecraft:stone", "minecraft:oak_planks"]`. Ids blockpedia has
        *no color for are silently skipped — check `len` afterwards.
        */
        fun fromBlockIds(idsJson: String): Result<Palette> {
            val idsJsonSliceMemory = PrimitiveArrayTools.borrowUtf8(idsJson)
            
            val returnVal = lib.Palette_from_block_ids(idsJsonSliceMemory.slice);
            try {
                val nativeOkVal = returnVal.getNativeOk();
                if (nativeOkVal != null) {
                    val selfEdges: List<Any> = listOf()
                    val handle = nativeOkVal 
                    val returnOpaque = Palette(handle, selfEdges, true)
                    return returnOpaque.ok()
                } else {
                    return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
                }
            } finally {
                idsJsonSliceMemory.close()
            }
        }
    }
    
    /** A copy of this palette ordered by perceptual lightness (Oklab L,
    *dark → light). Combined with `block_ids_json`, gives a
    *ready-to-index ramp: `ids[i]` for intensity `i / (len - 1)`.
    */
    fun sortedByLightness(): Palette {
        
        val returnVal = lib.Palette_sorted_by_lightness(handle);
        val selfEdges: List<Any> = listOf()
        val handle = returnVal 
        val returnOpaque = Palette(handle, selfEdges, true)
        return returnOpaque
    }
    
    /** JSON array of exactly `steps` DISTINCT block ids forming the
    *smoothest ramp this palette can make from (`r1`,`g1`,`b1`) to
    *(`r2`,`g2`,`b2`): targets are evenly spaced along the Oklab line
    *and blocks are chosen by a minimum-cost monotonic matching, so
    *off-hue blocks are penalized and no block repeats. Errors with
    *`InvalidArgument` when the palette has fewer than `steps` blocks,
    *`steps` is 0, or start equals end.
    */
    fun rampIdsJson(r1: UByte, g1: UByte, b1: UByte, r2: UByte, g2: UByte, b2: UByte, steps: UInt): Result<String> {
        val write = DW.lib.diplomat_buffer_write_create(0)
        val returnVal = lib.Palette_ramp_ids_json(handle, FFIUint8(r1), FFIUint8(g1), FFIUint8(b1), FFIUint8(r2), FFIUint8(g2), FFIUint8(b2), FFIUint32(steps), write);
        val nativeOkVal = returnVal.getNativeOk();
        if (nativeOkVal != null) {
            
            val returnString = DW.writeToString(write)
            return returnString.ok()
        } else {
            return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
        }
    }
    
    /** JSON array of exactly `steps` block ids sampling the color
    *gradient from (`r1`,`g1`,`b1`) to (`r2`,`g2`,`b2`) in Oklab
    *space, each step snapped to this palette's closest block. Built
    *for value→block lookups (heatmaps, fractals): index the returned
    *list by `intensity * (steps - 1)`. Entries may repeat on coarse
    *palettes; errors with `NotFound` on an empty palette.
    */
    fun gradientIdsJson(r1: UByte, g1: UByte, b1: UByte, r2: UByte, g2: UByte, b2: UByte, steps: UInt): Result<String> {
        val write = DW.lib.diplomat_buffer_write_create(0)
        val returnVal = lib.Palette_gradient_ids_json(handle, FFIUint8(r1), FFIUint8(g1), FFIUint8(b1), FFIUint8(r2), FFIUint8(g2), FFIUint8(b2), FFIUint32(steps), write);
        val nativeOkVal = returnVal.getNativeOk();
        if (nativeOkVal != null) {
            
            val returnString = DW.writeToString(write)
            return returnString.ok()
        } else {
            return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
        }
    }
    
    /** Number of blocks in the palette.
    */
    fun len(): ULong {
        
        val returnVal = lib.Palette_len(handle);
        return (returnVal.toULong())
    }
    
    /** The palette's block ids as a JSON array string.
    */
    fun blockIdsJson(): String {
        val write = DW.lib.diplomat_buffer_write_create(0)
        val returnVal = lib.Palette_block_ids_json(handle, write);
        
        val returnString = DW.writeToString(write)
        return returnString
    }
    
    /** The palette block whose color is closest (Oklab distance) to the
    *given RGB. Errors with `NotFound` on an empty palette.
    */
    fun closestBlock(r: UByte, g: UByte, b: UByte): Result<String> {
        val write = DW.lib.diplomat_buffer_write_create(0)
        val returnVal = lib.Palette_closest_block(handle, FFIUint8(r), FFIUint8(g), FFIUint8(b), write);
        val nativeOkVal = returnVal.getNativeOk();
        if (nativeOkVal != null) {
            
            val returnString = DW.writeToString(write)
            return returnString.ok()
        } else {
            return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
        }
    }

}